cmd/list.go

```diff
@@ -1,42 +1,178 @@
 package cmd
 
 import (
+	"context"
 	"fmt"
 	"os"
 	"sort"
+	"strings"
 	"text/tabwriter"
 
-	"github.com/whalebrew/whalebrew/packages"
+	"github.com/Masterminds/semver"
+	"github.com/docker/distribution/reference"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/filters"
+	"github.com/docker/go-units"
 	"github.com/spf13/cobra"
 	"github.com/spf13/viper"
+	"github.com/whalebrew/whalebrew/client"
+	"github.com/whalebrew/whalebrew/packages"
 )
 
 func init() {
 	RootCmd.AddCommand(listCommand)
 }
 
+type listInfo struct {
+	version      *semver.Version
+	imageSummary *types.ImageSummary
+	tag          string
+}
+
+type listInfos []*listInfo
+
+func (c listInfos) Len() int {
+	return len(c)
+}
+
+// Less is needed for the sort interface to compare two Version objects on the
+// slice. If checks if one is less than the other.
+func (c listInfos) Less(i, j int) bool {
+	// Sort tags that aren't semver to front,
+	// this shouldn't happen since we pre filter those to the back
+	if c[i].version == nil {
+		return false
+	} else if c[j].version == nil {
+		return true
+	} else {
+		return c[i].version.LessThan(c[j].version)
+	}
+}
+
+// Swap is needed for the sort interface to replace the Version objects
+// at two different positions in the slice.
+func (c listInfos) Swap(i, j int) {
+	c[i], c[j] = c[j], c[i]
+}
+
 var listCommand = &cobra.Command{
-	Use:   "list",
-	Short: "List installed packages",
+	Use:   "list [image]",
+	Short: "List installed packages, or installable images for a package",
 	RunE: func(cmd *cobra.Command, args []string) error {
-		pm := packages.NewPackageManager(viper.GetString("install_path"))
-		packages, err := pm.List()
-		if err != nil {
-			return err
+		ctx := context.Background()
+
+		if arglen := len(args); arglen > 0 {
+			if arglen > 1 {
+				return fmt.Errorf("invalid usage, this command accepts 0 or 1 arguments")
+			}
+
+			cli, err := client.NewClient()
+			if err != nil {
+				return err
+			}
+
+			return listImagesForPackage(ctx, cli, args[0])
+		} else {
+			return listInstalled()
 		}
+	},
+}
+
+func listImagesForPackage(ctx context.Context, cli *client.Client, pkgName string) error {
+	pm := packages.NewPackageManager(viper.GetString("install_path"))
+	if installed, err := pm.HasInstallation(pkgName); !installed {
+		return err
+	}
+
+	pkg, err := pm.Load(pkgName)
+	if err != nil {
+		return err
+	}
 
-		packageNames := make([]string, 0, len(packages))
-		for k := range packages {
-			packageNames = append(packageNames, k)
+	ref, err := reference.ParseAnyReference(pkg.Image)
+	if err != nil {
+		return err
+	}
+
+	installedRef, ok := ref.(reference.NamedTagged)
+	if !ok {
+		return nil
+	}
+
+	installedTag := installedRef.Tag()
+
+	name := strings.TrimPrefix(installedRef.Name(), "docker.io/library/")
+
+	summaries, err := cli.ImageList(ctx, types.ImageListOptions{
```

Comment:
I am wondering about rather the interface of using `list`
Using this integration would mean that images are already downloaded on the local machine, as a cache of previously installed versions, and are listed through `docker images <image>`
Would not it make more sense use `whalebrew search <term>`, and in case the image name matches an exact image list remote tags, leaving `whalebrew list` to consider packages, hence images rather than command names. When running `search`, you are having questions like "what is available", versus when you are running `list` you are trying to answer questions like "what do I have installed"

Reply:
Absolutely that seems like a much better way. I think it would be nice to include local cached images in the result in case an image is from a private registry. This would also enable specifying which tags exist locally

Reply:
Indeed, good idea, would a column like "DOWNLOADED" in the output be an idea?